src/TzBot/Util.hs

```
@@ -156,3 +160,22 @@ instance (FromJSON a) => FromJSON (WithUnknown a) where
 
 instance (ToJSON a) => ToJSON (WithUnknown a) where
   toJSON (WithUnknown eith) = either (String . show) toJSON eith
+
+postfixFields :: LensRules
+postfixFields = lensRules & lensField .~ mappingNamer (\n -> [n ++ "L"])
+
+----
+-- not present in mtl-2.2.2
+tryError :: MonadError e m => m a -> m (Either e a)
+tryError action = (Right <$> action) `catchError` (pure . Left)
+
+catchAllErrors
+  :: (MonadError e m, MonadBaseControl IO m)
```

I guess you use `MonadBaseControl` because `ExceptT` layer does not pass `MonadCatch`?

Hmm, there is a problem with this: we generally try to use exception utilities that don't catch asynchronous exceptions by default. E.g. `safe-exception` package, `unliftio` package with its `UnliftIO.Exception` module, and `universum` do this, while classic `Control.Exceptions`, `mtl`-style exceptions, and `Control.Exception.Lifted` don't.

Maybe there would be other easy way to workaround that, without using other `try`?

---

Well, in this function we really would like to catch all exceptions (and asynchronous ones), is it bad? We can also add a haddock, e.g. "This function catches all the exceptions including asynchronous ones"

---

I guess the best thing we can do here is to throw away ExceptT IO and just use MonadThrow, but maybe it's better to implement in another PR

---

> Well, in this function we really would like to catch all exceptions (and asynchronous ones),

If we try to abort the program then I suppose this thread should not catch the exception and report it as a problem, it is a normal situation.

> I guess the best thing we can do here is to throw away ExceptT IO and just use MonadThrow, but maybe it's better to implement in another PR

Agree, let's handle this separately.